config/v1/types_securityprofile.go

```diff
@@ -0,0 +1,316 @@
+	//     - ECDHE-ECDSA-AES128-GCM-SHA256
+	//     - ECDHE-RSA-AES128-GCM-SHA256
+	//     - ECDHE-ECDSA-AES256-SHA384
+	//     - ECDHE-RSA-AES256-SHA384
+	//     - ECDHE-ECDSA-AES128-SHA256
+	//     - ECDHE-RSA-AES128-SHA256
+	//   tlsVersion:
+	//     minimumVersion: TLSv1.2
+	//     maximumVersion: TLSv1.2
+	//   dhParamSize: 2048
+	//
+	// +optional
+	// +nullable
+	Modern *ModernTLSProfile `json:"modern,omitempty"`
+	// custom is a user-defined TLS security profile. Be extremely careful using a custom
+	// profile as invalid configurations can be catastrophic. An example custom profile
+	// looks like this:
+	//
+	//   ciphers:
+	//     - ECDHE-ECDSA-CHACHA20-POLY1305
+	//     - ECDHE-RSA-CHACHA20-POLY1305
+	//     - ECDHE-RSA-AES128-GCM-SHA256
+	//     - ECDHE-ECDSA-AES128-GCM-SHA256
+	//   tlsVersion:
+	//     minimumVersion: TLSv1.1
+	//     maximumVersion: TLSv1.2
+	//   dhParamSize: 1024
+	//
+	// +optional
+	// +nullable
+	Custom *CustomTLSProfile `json:"custom,omitempty"`
+}
+
+// OldTLSProfile is a TLS security profile based on:
+// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+type OldTLSProfile struct{}
+
+// IntermediateTLSProfile is a TLS security profile based on:
+// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+type IntermediateTLSProfile struct{}
+
+// ModernTLSProfile is a TLS security profile based on:
+// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+type ModernTLSProfile struct{}
+
+// CustomTLSProfile is a user-defined TLS security profile. Be extremely careful
+// using a custom TLS profile as invalid configurations can be catastrophic.
+type CustomTLSProfile struct {
+	TLSProfileSpec `json:",inline"`
+}
+
+// TLSProfileType defines a TLS security profile type.
+type TLSProfileType string
+
+const (
+	// Old is a TLS security profile based on:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+	TLSProfileOldType TLSProfileType = "Old"
+	// Intermediate is a TLS security profile based on:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	TLSProfileIntermediateType TLSProfileType = "Intermediate"
+	// Modern is a TLS security profile based on:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+	TLSProfileModernType TLSProfileType = "Modern"
+	// Custom is a TLS security profile that allows for user-defined parameters.
+	TLSProfileCustomType TLSProfileType = "Custom"
+)
+
+// TLSProfileSpec is the desired behavior of a TLSSecurityProfile.
+type TLSProfileSpec struct {
+	// ciphers is used to specify the cipher algorithms that are negotiated
+	// during the TLS handshake. Preface a cipher with a "!" to disable a
+	// specific cipher from being negotiated. Note that disabled ciphers must
+	// be quoted due to the leading "!". For example, to use 3DES but not
+	// EDH-DSS-DES-CBC3-SHA (yaml):
```

Comment:
Something I don't remember discussing in the past iterations... the `!` prefix support does seem like some sort of end-run around defining an expression or other API. For example, it would be possible to model this as:

```go
type TLSProfileSpec struct {
  EnabledCiphers []string
  DisabledCiphers []string
}
```

or...

```go
type TLSProfileSpec struct {
  Ciphers []TLSProfileCipher
}

type TLSProfileCipher struct {
  Cipher string
  // Enabled, Disabled, etc.
  Policy TLSCipherPolicy
}
```

And probably others. It may not matter in this case, but I wanted to at least call it out.

Reply:
+1 for
```
type TLSProfileSpec struct {
  EnabledCiphers []string
  DisabledCiphers []string
}
```

Reply:
We have gone with whitelist only so it is safe to skip anything unrecognized.